### PR TITLE
dyno: factor VarScopeVisitor out of split init implementation

### PR DESCRIPTION
**Pull request by mppf**
Follow-up to PR #20843.

As I am looking at implementing the copy elision analysis in dyno, I'm realizing that many of the details are the same as for split init. So, in this PR, I've factored out these details into VarScopeVisitor.h / VarScopeVisitor.cpp (which are private within the library). This visitor can have subclasses for split init / copy elision / etc. It should just be updated to handle whatever these similar cases need.

This is primarily a refactor. The only behavior change that I've added intentionally is that the replacement for `computeOutFormals`, which is `computeActualFormalIntents`, is now more picky about making sure that the return intent overloads have the same types and intents.

Reviewed by @DanilaFe - thanks!

### CODE REVIEW

File: frontend/lib/resolution/split-init.cpp [hunk 1]
```diff
@@ -19,196 +19,121 @@
...
+}
+void FindSplitInits::addDeclaration(VarFrame* frame,
+                                    const VarLikeDecl* ast) {
+  bool inserted = frame->addToDeclaredVars(ast->id());
+  if (inserted) {
+    if (ast->initExpression() == nullptr) {
+      frame->eligibleVars.insert(ast->id());
+    }
   }
 }
```
Comment (DanilaFe) [inline frontend/lib/resolution/split-init.cpp:120]: Implementing custom logic on `VarFrame`s in subclasses is a slight breeze in the direction of the "more type safe" approach (which specializes the VarFrames somehow per-use). However, all things considered, it's probably still better to continue with the simple approach.

File: frontend/lib/resolution/split-init.cpp [hunk 2]
```diff
@@ -218,27 +143,66 @@ void FindSplitInits::handleInitOrAssign(ID dstId,
...
 // the then/else blocks from the conditional
-void FindSplitInits::handleExitConditional(const uast::Conditional* cond) {
-  SplitInitFrame* frame = scopeStack.back().get();
+void FindSplitInits::handleConditional(const Conditional* cond) {
+  VarFrame* frame = currentFrame();
   assert(frame->subBlocks.size() == 1 || frame->subBlocks.size() == 2);
 
-  SplitInitFrame* thenFrame = frame->subBlocks[0].frame.get();
-  SplitInitFrame* elseFrame = nullptr;
+  VarFrame* thenFrame = frame->subBlocks[0].frame.get();
```
Comment (DanilaFe) [inline frontend/lib/resolution/split-init.cpp:204]: Since we have a `currentFrame()`, which not have a `subBlockFrame(i)` method?

**Review by DanilaFe — APPROVED** [review]
Looks good to me :)

File: frontend/lib/resolution/split-init.cpp (hunk 2 above)
Comment (mppf) [reply frontend/lib/resolution/split-init.cpp]: I did this one

File: frontend/lib/resolution/VarScopeVisitor.cpp [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+void
+computeActualFormalIntents(Context* context,
+                           const MostSpecificCandidates& candidates,
+                           const CallInfo& ci,
+                           const std::vector<const AstNode*>& actualAsts,
+                           std::vector<IntentList>& actualFormalIntents,
+                           std::vector<QualifiedType>& actualFormalTypes) {
+
+  int nActuals = ci.numActuals();
+  actualFormalIntents.clear();
```
[the remark this answers is not in the document]
Comment (mppf) [reply frontend/lib/resolution/VarScopeVisitor.cpp:392]: I think I'll leave it as-is but we could change that in the future if we have a reason to.

File: frontend/lib/resolution/VarScopeVisitor.cpp [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+    for (const TypedFnSignature* fn : candidates) {
+      if (fn != nullptr) {
+        int n = fn->numFormals();
+        for (int i = 0; i < n; i++) {
+          const QualifiedType& formalQt = fn->formalType(i);
+          auto kind = formalQt.kind();
+          if (kind == QualifiedType::OUT ||
+              kind == QualifiedType::IN ||
+              kind == QualifiedType::CONST_IN ||
+              kind == QualifiedType::INOUT) {
```
[the remark this answers is not in the document]
Comment (mppf) [reply frontend/lib/resolution/VarScopeVisitor.cpp]: Those are interesting but right now I'd like to keep the simpler code here. Part of the reason is that I know this code will need to handle these intents but perhaps in the future it needs to handle others. So I have not been thinking it inherent to the way 'in' / 'const in' / 'out' / 'inout' have something in common; rather it's just what the code working with this helper will need.

**Pull request merged**